### PR TITLE
DR2-1785 Add files change handler.

### PR DESCRIPTION
**Pull request by MancunianSam**
There's a load of changes linked to the restructure of the
da-aws-clients classes as well.

### CODE REVIEW

File: ingest-files-change-handler/src/test/scala/uk/gov/nationalarchives/ingestfileschangehandler/Utils.scala [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+              case Type.File => Option(row.createFile().asInstanceOf[T])
+              case _ => None
+          })
+
+      }
+
+      override def updateAttributeValues(dynamoDbRequest: DADynamoDBClient.DADynamoDbRequest): IO[Int] = IO(1)
+
+      override def queryItems[U](tableName: String, requestCondition: RequestCondition, potentialGsiName: Option[String] = None)(using returnTypeFormat: DynamoFormat[U]): IO[List[U]] = ref.get.map { rows =>
+        if(potentialGsiName.isEmpty) {
```
Comment (techncl) [inline ingest-files-change-handler/src/test/scala/uk/gov/nationalarchives/ingestfileschangehandler/Utils.scala]: You have repetition here and quite a few braces and brackets, could you not do this:

```
        (for {
          values <- requestCondition.dynamoValues
          map <- values.toMap[String].toOption
        } yield
          if(potentialGsiName.isEmpty)
              rows.filter(row => map.get("conditionAttributeValue0").contains(row.id.toString))
              .map(_.createAsset().asInstanceOf[U])
          else
            rows.filter(row => row.parentPath == map.get("parentPath") && map.get("batchId").contains(row.batchId))
            .map(_.createFile().asInstanceOf[U])
        ).getOrElse(Nil)
```
or 

```
        (for {
          values <- requestCondition.dynamoValues
          map <- values.toMap[String].toOption
        } yield {
          if(potentialGsiName.isEmpty)
            rows.filter(row => map.get("conditionAttributeValue0").contains(row.id.toString))
          else
            rows.filter(row => row.parentPath == map.get("parentPath") && map.get("batchId").contains(row.batchId))
        }.map(_.createAsset().asInstanceOf[U])
        ).getOrElse(Nil)
```

**Review by techncl — COMMENTED** [review]
Looks good! A few comments here albeit, relatively minor, as well as:

The `handler` method is quite large; could some of the methods inside it live outside it as private methods?

File: dynamo-formatters/src/main/scala/uk/gov/nationalarchives/dynamoformatters/DynamoFormatters.scala [hunk 2]
```diff
@@ -106,7 +107,17 @@ object DynamoFormatters {
 
   given lockTablePkFormat: Typeclass[LockTablePartitionKey] = deriveDynamoFormat[LockTablePartitionKey]
 
+  given typeFormatter: DynamoFormat[Type] = new DynamoFormat[Type]:
+    override def read(av: DynamoValue): Either[DynamoReadError, Type] = av.as[String].map(Type.valueOf)
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply dynamo-formatters/src/main/scala/uk/gov/nationalarchives/dynamoformatters/DynamoFormatters.scala:111]: It's the name in the `DynamoFormat` trait so it gets added when Intellij overrides the methods. I've changed it to dynamoValue

File: dynamo-formatters/src/test/scala/uk/gov/nationalarchives/dynamoformatters/DynamoFormattersTest.scala [hunk 3]
```diff
@@ -328,6 +331,19 @@ class DynamoFormattersTest extends AnyFlatSpec with TableDrivenPropertyChecks wi
     assetRowInvalidValue.ingestedPreservica should equal(false)
   }
 
+  "assetTableFormat read" should "return true for ingested_CC if the value is true and false otherwise" in {
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply dynamo-formatters/src/test/scala/uk/gov/nationalarchives/dynamoformatters/DynamoFormattersTest.scala:334]: Good point, I've added a check for that.

File: ingest-files-change-handler/README.md [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+## Environment Variables
+
+| Name              | Description                                                               |
+|-------------------|---------------------------------------------------------------------------|
+| DYNAMO_TABLE_NAME | The name of the ingest files dynamo table                                 |
+| DYNAMO_GSI_NAME   | The global secondary index name. Used to search by batchId and parentPath |
+| TOPIC_ARN         | The output SNS topic arn                                                  |
+
+## Lambda logic
+The logic which determines whether to send a message and which message is described with this pseudocode
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply ingest-files-change-handler/README.md:67]: Good spot, there is, thanks.

File: ingest-files-change-handler/TestCases.md [hunk 5]
```diff
@@ -0,0 +1,813 @@
...
+| id | batchId | type          | parentPath | ingested_PS |
+|----|---------|---------------|------------|-------------|
+| 0  | A       | ArchiveFolder |            |             |
+| 1  | A       | Asset         | 0/         | true        |
+| 2  | A       | File          | 0/1/       |             |
+| 3  | A       | File          | 0/1/       |             |
+
+### Input
+
+```JSON
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply ingest-files-change-handler/TestCases.md:21]: They're not the same for each test case. The NewImage object changes for quite a few of them.

File: ingest-files-change-handler/TestCases.md [hunk 6]
```diff
@@ -0,0 +1,813 @@
...
+    }
+  ]
+}
+```
+
+### Output (messages sent)
+
+No output.
+
+## 7. `ingested_CC` for 1 File only where not all files are complete
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply ingest-files-change-handler/TestCases.md:472]: 6 is an update message for an asset row and 7 is an update for a file row.

File: ingest-files-change-handler/src/main/scala/uk/gov/nationalarchives/ingestfileschangehandler/Lambda.scala [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+  override def handler: (DynamodbEvent, Config, Dependencies) => IO[Unit] = (event, config, dependencies) => {
+    def getPrimaryKey[T <: DynamoTable](row: T) =
+      FilesTablePrimaryKey(FilesTablePartitionKey(row.id), FilesTableSortKey(row.batchId))
+
+    def sendOutputMessage(asset: AssetDynamoTable, messageType: MessageType): IO[Unit] = {
+      val message = OutputMessage(OutputProperties(dependencies.uuidGenerator(), None, dependencies.instantGenerator(), messageType), OutputParameters(asset.id))
+      dependencies.daSnsClient.publish(config.topicArn)(message :: Nil).map(_ => ())
+    }
+
+    def completeAsset(asset: AssetDynamoTable, children: List[FileDynamoTable]) = {
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply ingest-files-change-handler/src/main/scala/uk/gov/nationalarchives/ingestfileschangehandler/Lambda.scala:34]: It's a fair point, I've split them into two, `sendMessageAndDelete` and `sendMessageAndDeleteSkippedAssets`

File: ingest-files-change-handler/src/main/scala/uk/gov/nationalarchives/ingestfileschangehandler/Lambda.scala [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+    }
+
+    def processFile(fileRow: FileDynamoTable): IO[Unit] = for {
+      parentId <- IO.fromOption {
+        fileRow.parentPath
+          .flatMap(_.split('/').lastOption)
+          .map(UUID.fromString)
+      }(new Exception(s"Cannot find a direct parent for file ${fileRow.id}"))
+      primaryKey <- IO.pure(FilesTablePrimaryKey(FilesTablePartitionKey(parentId), FilesTableSortKey(fileRow.batchId)))
+      parentAssetList <- dependencies.daDynamoDbClient.getItems[AssetDynamoTable, FilesTablePrimaryKey](List(primaryKey), config.dynamoDbTable)
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply ingest-files-change-handler/src/main/scala/uk/gov/nationalarchives/ingestfileschangehandler/Lambda.scala]: parentAssets is ok.

File: ingest-files-change-handler/src/main/scala/uk/gov/nationalarchives/ingestfileschangehandler/Lambda.scala [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+      dependencies.daSnsClient.publish(config.topicArn)(message :: Nil).map(_ => ())
+    }
+
+    def completeAsset(asset: AssetDynamoTable, children: List[FileDynamoTable]) = {
+      for {
+        _ <- sendOutputMessage(asset, IngestComplete)
+        childKeys <- IO.pure(children.map(getPrimaryKey))
+        assetKeys <- IO.pure(List(asset).map(getPrimaryKey))
+        _ <- dependencies.daDynamoDbClient.deleteItems(config.dynamoDbTable, childKeys ++ assetKeys)
+        assetsSameId <- dependencies.daDynamoDbClient.queryItems[AssetDynamoTable](config.dynamoDbTable, "id" === asset.id.toString)
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply ingest-files-change-handler/src/main/scala/uk/gov/nationalarchives/ingestfileschangehandler/Lambda.scala]: It is a method, it's the queryItems method. Or do you mean like)
```scala
def assetsSameId(asset: AssetDynamoTable): IO[List[AssetDynamoTable]
  dependencies.daDynamoDbClient.queryItems[AssetDynamoTable](config.dynamoTableName, "id" === asset.id.toString)

I don't think that saves us much in terms of characters. I'll leave it.

File: ingest-files-change-handler/src/main/scala/uk/gov/nationalarchives/ingestfileschangehandler/Lambda.scala [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+      for {
+        children <- dependencies.daDynamoDbClient
+          .queryItems[FileDynamoTable](config.dynamoDbTable, "batchId" === asset.batchId and "parentPath" === childrenParentPath, Option(config.gsiName))
+        _ <- IO.raiseWhen(children.length != asset.childCount)(
+          new Exception(s"Asset id ${asset.id}: has ${asset.childCount} children in the files table but found ${children.length} children in the Preservation system")
+        )
+      } yield children
+    }
+
+    def processFile(fileRow: FileDynamoTable): IO[Unit] = for {
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply ingest-files-change-handler/src/main/scala/uk/gov/nationalarchives/ingestfileschangehandler/Lambda.scala:58]: I like this. The point about it being more modular is a good one. I've changed it to `getParentAsset` and then the match is now
```scala
case fileRow: FileDynamoTable   => getParentAsset(fileRow).flatMap(processAsset)
```

File: ingest-files-change-handler/src/test/scala/uk/gov/nationalarchives/ingestfileschangehandler/LambdaTest.scala [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+  val folderB: DynamoRow = DynamoRow(UUID.randomUUID, "B", ArchiveFolder, None)
+  val assetB: DynamoRow = DynamoRow(assetA.id, "B", Asset, Option(s"${folderB.id}"), true)
+  val fileBOne: DynamoRow = DynamoRow(UUID.randomUUID, "B", File, Option(s"${folderB.id}/${assetB.id}"))
+  val fileBTwo: DynamoRow = DynamoRow(UUID.randomUUID, "B", File, Option(s"${folderB.id}/${assetB.id}"))
+
+  def outputBuilder(id: UUID, messageType: MessageType): OutputMessage = OutputMessage(OutputProperties(messageId, None, instant, messageType), OutputParameters(id))
+
+  val handlerOutputsTable: TableFor4[String, List[DynamoRow], DynamoRow, List[OutputMessage]] = Table(
+    ("title", "rowsInTable", "newRowInput", "expectedOutput"),
+    ("ingested_PS for 1 item only sends an update message",
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply ingest-files-change-handler/src/test/scala/uk/gov/nationalarchives/ingestfileschangehandler/LambdaTest.scala]: It's item as in a DynamoDB item I think so I'll leave that. I've added the comma in and a comma in all the other test titles.

File: ingest-files-change-handler/src/test/scala/uk/gov/nationalarchives/ingestfileschangehandler/LambdaTest.scala [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+        fileATwo
+      ), assetA, List(outputBuilder(assetA.id, IngestUpdate))),
+    ("skipIngest for 1 item only sends no message",
+      List(
+        folderA,
+        assetA.copy(skipIngest = true, ingestedPreservica = false),
+        fileAOne,
+        fileATwo
+      ), assetA.copy(skipIngest = true, ingestedPreservica = false), Nil),
+    ("ingested_PS+skipIngest for 1 item only sends ingest complete message",
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply ingest-files-change-handler/src/test/scala/uk/gov/nationalarchives/ingestfileschangehandler/LambdaTest.scala]: I've replaced it with the word "and"

File: ingest-files-change-handler/src/test/scala/uk/gov/nationalarchives/ingestfileschangehandler/Utils.scala (hunk 1 above)
Comment (MancunianSam) [reply ingest-files-change-handler/src/test/scala/uk/gov/nationalarchives/ingestfileschangehandler/Utils.scala]: The first one is good, I've changed it.

**Comment by MancunianSam** [comment]
I think I'll leave the handler method. It is separated into methods inside the handler method. This is mostly so I don't have to pass in the Dependencies and Config objects to each method. I could put it in a seperate class but I don't know if it's worth it. We can always change it later.

File: ingest-files-change-handler/TestCases.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (techncl) [reply ingest-files-change-handler/TestCases.md:21]: Ah, yes! Sorry, I must have been tired when comparing them.

**Review by techncl — APPROVED** [review]
👍

**Pull request merged**